satpy/readers/viirs_sdr.py

```diff
@@ -219,18 +219,29 @@ def get_file_units(self, dataset_id, ds_info):
             LOG.debug("Unknown units for file key '%s'", dataset_id)
         return file_units
 
-    def scale_swath_data(self, data, scaling_factors):
+    def scale_swath_data(self, data, scaling_factors, dataset_group):
         """Scale swath data using scaling factors and offsets.
 
         Multi-granule (a.k.a. aggregated) files will have more than the usual two values.
         """
-        num_grans = len(scaling_factors) // 2
-        gran_size = data.shape[0] // num_grans
+        rows_per_gran = self._get_rows_per_granule(dataset_group)
         factors = scaling_factors.where(scaling_factors > -999, np.float32(np.nan))
-        factors = factors.data.reshape((-1, 2))
-        factors = xr.DataArray(da.repeat(factors, gran_size, axis=0),
-                               dims=(data.dims[0], 'factors'))
-        data = data * factors[:, 0] + factors[:, 1]
+        factors = factors.data.reshape((-1, 2)).rechunk((1, 2))  # make it so map_blocks happens per factor
```

Comment:
The abstraction level of these two lines do not match the two other lines, should these be integrated into `_map_and_apply_factors`, or be extracted into a separate method?

Reply:
Done by extracting to another method. Maybe not the best overall as now it is a method that is just calling three other methods but...I guess it is better.